src/dune_engine/build.ml

```diff
@@ -353,79 +359,116 @@ let fold_labeled (type acc) t ~(init : acc) ~f =
-
-    let id m = m.id
+let do_not_use_stage_fiber f = Fiber f
+let do_not_use_stage_dyn_fiber f = Dyn_fiber f
+let do_not_use_stage_build f = Build f
+
+module Make_exec (Build_deps:sig val build_deps: Dep.Set.t -> unit Fiber.t end) = struct
+
+  module rec Execution : sig
+    val exec : 'a t -> ('a * Dep.Set.t) Fiber.t
+  end = struct
+    module Function = struct
+      type 'a input = 'a memo
+
+      type 'a output = 'a * Dep.Set.t
+
+      let name = "exec-memo"
+
+      let id m = m.id
+
+      let to_dyn m = Dyn.String m.name
+
+      let eval m = Execution.exec m.t
+    end
+
+    module Memo = Memo.Poly.Async (Function)
+
+    let exec t =
+      let open Fiber.O in
+      let file_exists = Fdecl.get file_exists_fdecl in
+      let eval_pred = Fdecl.get eval_pred_fdecl in
+      let rec go : type a. a t -> (a * Dep.Set.t) Fiber.t =
+        fun t ->
+          match t with
+          | Pure x -> Fiber.return (x, Dep.Set.empty)
+          | Map (f, a) ->
+            let+ a, dyn_deps_a = go a in
+            (f a, dyn_deps_a)
+          | Both (a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            ((a, b), Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Seq (a, b) ->
+            let* (), dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Map2 (f, a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (f a b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Deps _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_for_rule _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_glob g -> Fiber.return ((eval_pred g : Path.Set.t), Dep.Set.empty)
+          | Contents p -> Fiber.return (Io.read_file p, Dep.Set.empty)
+          | Lines_of p -> Fiber.return (Io.lines_of_file p, Dep.Set.empty)
+          | Dyn_paths t ->
+            let+ (x, paths), dyn_deps = go t in
+            (x, Dep.Set.add_paths dyn_deps paths)
+          | Dyn_deps t ->
+            let+ (x, dyn_deps), dyn_deps_x = go t in
+            (x, Dep.Set.union dyn_deps dyn_deps_x)
+          | Label _ -> Fiber.return ((), Dep.Set.empty)
+          | Fail { fail } -> fail ()
+          | If_file_exists (p, then_, else_) ->
+            if file_exists p then
+              go then_
+            else
+              go else_
+          | Filter_existing_files p ->
+            let+ (x, files), dyn_deps = go p in
+            let files = Path.Set.filter ~f:file_exists files in
+            ((x, files), dyn_deps)
+          | Catch (t, on_error) -> (
+              let+ res = Fiber.fold_errors
+                           ~init:None
+                           ~on_error:(fun bt _ -> Some (on_error bt.Exn_with_backtrace.exn))
+                           (fun () -> go t)
+              in
+              match res with
+              | Ok r -> r
+              | Error None -> assert false
```

Comment:
Is this `assert false` just a consequence of poor interface of `Fiber.fold_errors` that makes all call sites deal with the impossible `Error None` case? Or is there a more interesting reason that this case never happens?

Reply:
It's not obvious to me that order in which errors are reported is preserved here. Can you write some comments explaining the semantics of this new `exec` implementation with respect to error-handling?

Reply:
OK, I've now mostly convinced myself that it's impossible to avoid the `assert false` using the current `Fiber.fold_errors` API.

There is still the question about the order in which errors are reported: does anything guarantee that the order is the same? Or can it be different?

Reply:
Huh, I'm surprised that there is only one use site of `catch` and that it's so primitive :)

Your change seems fine to me, although I'd have a hard time guessing what `don't_fail` should do without looking at its type. Keeping the original name `catch` seems better to me.